R/zchunk_L221.en_supply.R

```diff
@@ -92,295 +105,661 @@ module_energy_L221.en_supply_DISABLED <- function(command, ...) {
+      ungroup() %>%
+      filter(year %in% MODEL_YEARS) %>%
+      select(sector.name = supplysector, subsector.name = subsector, technology, minicam.non.energy.input, year, input.cost) -> L221.GlobalTechCost_en
+
+    # Shareweights of global technologies
+    A21.globaltech_shrwt %>%
+      gather(year, share.weight, matches(YEAR_PATTERN)) %>%
+      mutate(year = as.integer(year)) -> A21.globaltech_shrwt
+
+    A21.globaltech_shrwt %>%
+      select(supplysector, subsector, technology) %>%
+      distinct %>%
+      # Interpolate to all years
+      repeat_add_columns(tibble(year = c(HISTORICAL_YEARS, FUTURE_YEARS))) %>%
+      left_join(A21.globaltech_shrwt, by = c("supplysector", "subsector", "technology", "year")) %>%
+      group_by(supplysector, subsector, technology) %>%
+      mutate(share.weight = approx_fun(year, value = share.weight, rule = 1)) %>%
+      ungroup() %>%
+      filter(year %in% MODEL_YEARS) %>%
+      select(sector.name = supplysector, subsector.name = subsector, technology, year, share.weight) -> L221.GlobalTechShrwt_en
+
+    # Keywords of global technologies
+    A21.globaltech_keyword %>%
+      repeat_add_columns(tibble(year = c(HISTORICAL_YEARS, FUTURE_YEARS))) %>%
+      select(sector.name = supplysector, subsector.name = subsector, technology, primary.consumption, year) %>%
+      filter(year %in% MODEL_YEARS) -> L221.PrimaryConsKeyword_en
+
+    # Secondary feed outputs of biofuel production technologies
+    # NOTE: secondary outputs are only considered in future time periods
+    # NOTE: secondary outputs are only written for the regions/technologies where applicable, so the global tech database can not be used
+    # to get the appropriate region/tech combinations written out, first repeat by all regions, then subset as appropriate
+    A21.globaltech_secout %>%
+      repeat_add_columns(tibble(region = GCAM_region_names$region)) %>%
+      left_join(calibrated_techs %>%
+                  select(minicam.energy.input, sector), by = c("supplysector" = "minicam.energy.input")) -> L221.globaltech_secout_R
+
+    # For corn ethanol, region + sector is checked. For biodiesel, need to also include the region-specific feedstocks (to exclude palm oil biodiesel producing regions)
+    L221.globaltech_secout_R %>%
+      left_join(A_regions %>%
+                  select(region, ethanol, GCAM_region_ID) %>%
+                  rename(sector = ethanol), by = c("region", "sector")) %>%
+      left_join(A_regions %>%
+                  select(region, biodiesel, biomassOil_tech, GCAM_region_ID) %>%
+                  rename(sector = biodiesel, technology = biomassOil_tech), by = c("region", "sector", "technology")) %>%
+      filter(!is.na(GCAM_region_ID.x) | !is.na(GCAM_region_ID.y)) %>%
+      mutate(GCAM_region_ID = if_else(!is.na(GCAM_region_ID.x), GCAM_region_ID.x, GCAM_region_ID.y)) %>%
+      select(-GCAM_region_ID.x, -GCAM_region_ID.y) -> L221.globaltech_secout_R
```

Comment:
This is a pretty involved pipeline, with nested joins, multiple region IDs (??), etc. Glad it works but maybe expand the commenting in here a bit to clarify what's going on.